### PR TITLE
Only build common protos with --grpc option for LRO.

### PR DESCRIPTION
**Pull request by bjwatson**
Updates https://github.com/GoogleCloudPlatform/google-cloud-python/issues/1893#issuecomment-245366078

https://github.com/grpc/grpc/issues/8056 is a proposal to fix this issue in gRPC (FYI @jmuk)

### CODE REVIEW

**Comment by bjwatson** [comment]
@jmuk @garrettjonesgoogle @jskeet @vchudnov-g @pongad, do any of your languages have standalone `googleapis-common-protos` packages? If so, and if you're adding LRO service methods to it, please check in with me. I've been through a few battles with it already in Python, which might be relevant to you.

File: lib/api_repo.js [hunk 1]
```diff
@@ -1018,11 +1021,14 @@ ApiRepo.prototype._makeProtocFunc = function _makeProtocFunc(opts, language) {
           pluginOption = '--plugin=';
         } else {
           args.push('--' + language + '_out=' + outDir);
-          if (language === 'python') {
-            // required per https://github.com/grpc/grpc/issues/7857
-            args.push('--grpc_python_out=' + outDir);
-          } else {
-            args.push('--grpc_out=' + outDir);
+          if (!opts.buildCommonProtos ||
+              _.includes(COMMON_GRPC_PROTOS, protoPath)) {
```
Comment (jmuk) [inline lib/api_repo.js]: isn't this the path from the proto's rootdir (i.e. 'google/longrunning/operations.proto')?

File: lib/api_repo.js [hunk 2]
```diff
@@ -56,6 +56,9 @@ var DEFAULT_PROTO_COMPILER = 'protoc';
 // Python builds using grpcio-tools.
 var NO_PROTOC_PLUGIN = ['nodejs', 'java', 'python'];
 
+// common protos for which we build gRPC service methods
+var COMMON_GRPC_PROTOS = ['operations.proto'];
```
Comment (jmuk) [inline lib/api_repo.js]: Adding bytestream.proto as well?  It's not in the common protos right now, but might be worth adding here.

**Comment by jmuk** [comment]
This looks a bit weird, could be a bug of grpc tools for Python. Are there any bugs in grpc?
Might be worth filing a new one if not, and noted the link as a comment.

**Comment by bjwatson** [comment]
@jmuk Does gRPC codegen make the grpc imports conditional on actually using gRPC for other languages?

File: lib/api_repo.js (hunk 1 above)
Comment (bjwatson) [inline lib/api_repo.js]: No, it's just the last element. `pkg` has the path.

File: lib/api_repo.js (hunk 2 above)
Comment (bjwatson) [inline lib/api_repo.js]: Ok. I don't have the context on that, but I don't object.

**Comment by bjwatson** [comment]
@jmuk I checked Ruby and saw that it generates the message and services in different files, which avoids this problem. @garrettjonesgoogle told me the same thing for Java, thanks to their rigid one-class-per-file policy.

**Comment by jmuk** [comment]
That's right (although I didn't know what Java does). If Python generator needs to generate single file containing both message and service definition -- then, Python generator should add grpc import line conditionally.
To be clear, I'm not against this PR, this PR will be good under the current situation.  I'm saying this situation would be better to be recognized by the grpc-python team too.

**Comment by jskeet** [comment]
@bjwatson: Our "common types" (Google.Api.CommonProtos) package includes the LRO definition, but not any gRPC code for it. I'd anticipate putting that in a separate package which then depends on Google.Api.CommonProtos. At the moment, I don't think any of these packages are produced by the pipeline, although we may well want to change this at some point...

**Comment by garrettjonesgoogle** [comment]
@bjwatson , java has google-grpc-common-protos, which is including the Operations Grpc generated code now.

**Comment by bjwatson** [comment]
@garrettjonesgoogle So do you have two packages for the Java common protos: one with just the messages, and one with the full gRPC build? Where can I find them?

**Comment by bjwatson** [comment]
@jskeet A potential problem with that approach is if the `Google.Api.CommonProtos` already has the LRO messages (like the Python `googleapis-common-protos` package does), and then the gRPC generated code attempts to duplicate the messages in another package that depends on `Google.Api.CommonProtos` (the Python gRPC compiler generates both messages and service methods in the same file for a given .proto).

**Comment by jskeet** [comment]
The gRPC proto plugin only generates the services - we should be fine.

**Comment by garrettjonesgoogle** [comment]
@bjwatson , we have a single package with both the common protos and the gRPC-generated class for Operations. http://search.maven.org/#artifactdetails%7Ccom.google.api.grpc%7Cgrpc-google-common-protos%7C0.0.9%7Cjar

**Comment by bjwatson** [comment]
@garrettjonesgoogle So [googleapis-common-protos](http://search.maven.org/#artifactdetails%7Ccom.google.api.grpc%7Cgoogleapis-common-protos%7C0.0.3%7Cjar) has been abandoned?

I assume you don't have any concerns about adding a gRPC dependency, because it works fine with the Java version of GAE Standard, right?

**Comment by garrettjonesgoogle** [comment]
@bjwatson Yes, the java repo [googleapis-common-protos](http://search.maven.org/#artifactdetails%7Ccom.google.api.grpc%7Cgoogleapis-common-protos%7C0.0.3%7Cjar) is abandoned now.

I haven't tried this with GAE standard; theoretically it shouldn't cause an issue to have dependencies that are unused (although if someone tried to use the gRPC dependency directly, I think it wouldn't work).

**Comment by bjwatson** [comment]
@garrettjonesgoogle I was having problems with our gcloud-python users on GAE Standard (e.g. https://github.com/GoogleCloudPlatform/google-cloud-python/issues/1893#issuecomment-244444346), so I'm having to do more to make gRPC a soft dependency in the Python version of [googleapis-common-protos](https://pypi.python.org/pypi/googleapis-common-protos).

Hopefully you don't run into the same problems. Has `gcloud-java` done a release yet that uses your new package? I ran into my issue promptly since I released a new version of the same package, and `gcloud-python` pulled it in.

**Comment by garrettjonesgoogle** [comment]
@bjwatson `google-cloud-java` has had an indirect dependency on grpc through gax-java since April: https://github.com/GoogleCloudPlatform/google-cloud-java/commit/2cd4b0cdc1a87b662b68490bf52c98504dcf91a7 - so, I'm guessing it hasn't been a problem.

**Comment by jskeet** [comment]
Including the gRPC dependency in C# would be less desirable, I think - it's a large package due to the native binaries involved, so anywhere you don't need it, you really don't want it. That said, if all the other platforms take that approach, there are benefits in consistency, and there are relatively few places you'd want the common protos *without* gRPC at the moment...

**Comment by bjwatson** [comment]
@jskeet The primary driver right now for this in Python and other languages is supporting Speech v1beta. The [AsyncRecognize method](https://github.com/googleapis/googleapis/blob/master/google/cloud/speech/v1beta1/cloud_speech.proto#L39) is the primary method for the API that we need to support, and it returns an LRO object. Our gcloud counterparts need the gRPC service methods for LRO to make this work.

There will probably also be other APIs that drive the need for LRO.

**Comment by bjwatson** [comment]
@garrettjonesgoogle That's good to hear. google-cloud-python has always had a [soft dependency](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/e407bfb9ab5a0a43186668b9c1b30ae835ff486f/setup.py#L22) on gax-python, to avoid pulling in gRPC for HTTP1.1-only clients. Seems like Java doesn't have that requirement.

**Comment by jskeet** [comment]
@bjwatson: Yes, we need LRO in C# as well. The options are:

- Include the LRO and IAM gRPC APIs within Google.Api.CommonProtos (required Grpc.Core dependency in Google.Api.CommonProtos).
- Have separate Google.Longrunning and Google.Iam.V1 packages

We definitely need it one way or another; it's just which route we go down. But maybe that would be better discussion away from this PR.

**Comment by bjwatson** [comment]
@jskeet Sounds good. We've been wrestling with both LRO and IAM in the other languages, too. Your choice of venue to continue this conversation. :)

**Comment by bjwatson** [comment]
@jmuk @geigerj I made your suggested changes, and fixed the code coverage error. PTAL

File: lib/api_repo.js [hunk 3]
```diff
@@ -1018,11 +1021,14 @@ ApiRepo.prototype._makeProtocFunc = function _makeProtocFunc(opts, language) {
...
-          if (language === 'python') {
-            // required per https://github.com/grpc/grpc/issues/7857
-            args.push('--grpc_python_out=' + outDir);
-          } else {
-            args.push('--grpc_out=' + outDir);
+          if (!opts.buildCommonProtos ||
+              _.includes(COMMON_GRPC_PROTOS, protoPath)) {
+            if (language === 'python') {
+              // required per https://github.com/grpc/grpc/issues/7857
+              args.push('--grpc_python_out=' + outDir);
```
Comment (geigerj) [inline lib/api_repo.js]: I thought we decided not to do this in order to support GAE?

My understanding was that this required a more complex change to generate both gRPC and gRPC-less modules.

File: test/fixtures/fake_protoc [hunk 4]
```diff
@@ -26,7 +26,7 @@ main() {
     # fix out_dir in case protoc was run for the go plugin
     local out_dir=$(echo "$out_dir" | sed 's/.*--go_out=plugins=grpc:\([^ ]*\).*/\1/')
     # fix out_dir in case protoc was run for the python plugin
-    local out_dir=$(echo "$out_dir" | sed 's/.*--grpc_python_out=\([^ ]*\).*/\1/')
+    local out_dir=$(echo "$out_dir" | sed 's/.*--python_out=\([^ ]*\).*/\1/')
```
Comment (jmuk) [inline test/fixtures/fake_protoc]: nit: out_dir is already declared as `local` previously, I think you don't have to declare it again, just update it.

**Comment by jmuk** [comment]
LGTM, with a nit.

File: lib/api_repo.js (hunk 3 above)
Comment (geigerj) [inline lib/api_repo.js]: After offline discussion, I understand that this is actually an improvement over the current state of master, where every common-proto is being built with the gRPC plugin (and hence potentially GAE-incompatible).

**Comment by geigerj** [comment]
LGTM, although would like to get to a state without hand-edits as soon as we can to reduce the risk of accidents when regenerating the package.

**Comment by bjwatson** [comment]
@geigerj Agreed. https://github.com/googleapis/packman/issues/112 covers this, and it would be a good idea to fix it before we do our next release of common-protos.

**Comment by bjwatson** [comment]
@jmuk I fixed your nit. PTAL

**Comment by jmuk** [comment]
LGTM

**Pull request merged**